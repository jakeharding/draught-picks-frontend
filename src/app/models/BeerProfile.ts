/**
 * BeerProfile.ts
 *
 * Created by jake
 * Created on 2/27/18
 *
 * Model beer profiles.
 */


export default class BeerProfile {
  uuid: string;
  abv_low: string;
  abv_hi: string;
  ibu_low: string;
  ibu_hi: string;
  like_description: string;
  user: string;

  constructor() {

  }
}
