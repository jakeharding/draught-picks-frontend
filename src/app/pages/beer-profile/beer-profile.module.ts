import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { IonicModule } from '@ionic/angular';
import { BeerProfilePage } from './beer-profile.page';
import { ComponentsModule } from '../../components/components.module';
import { AutoCompleteModule } from 'ionic4-auto-complete';

@NgModule({
  imports: [
    IonicModule,
    RouterModule.forChild([
      {
        path: 'beer-profile',
        component: BeerProfilePage
      },
    ]),
    ComponentsModule,
    ReactiveFormsModule,
    CommonModule,
    FormsModule,
    AutoCompleteModule
  ],
  declarations: [BeerProfilePage]
})
export class BeerProfilePageModule {}
