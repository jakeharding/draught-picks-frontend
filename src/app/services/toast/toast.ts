import { Injectable } from '@angular/core';
import { ToastController } from '@ionic/angular';

/*
  Generated class for the ToastProvider provider.

  See https://angular.io/guide/dependency-injection for more info on providers
  and Angular DI.
*/
@Injectable({
  providedIn: 'root',
})
export class ToastProvider {
  static readonly defaultErrorMsg = 'We are having trouble connecting at the moment. ' +
  'Please check your network connection and try again.';
  private readonly errorToastClass = 'error-toast';
  private readonly successToastClass = 'success-toast';

  constructor(private toastCtrl: ToastController) {}

  errorToast(message: string = ToastProvider.defaultErrorMsg): void {
    this.createToast(message, this.errorToastClass);
  }

  successToast(message: string): void {
    this.createToast(message, this.successToastClass);
  }

  async createToast(message: string,
                    cssClass: string,
                    duration: number = 3000,
                    position: 'top' | 'bottom' | 'middle' = 'top'
              ) {
    const toast = await this.toastCtrl.create({ message, cssClass, duration, position });
    await toast.present();
  }

}
