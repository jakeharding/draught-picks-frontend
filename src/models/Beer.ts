/**
 * Beer.ts
 *
 * Created by jake
 * Created on 2/27/18
 *
 * Model the beers
 */
import BeerRating from "./BeerRating";


export default interface Beer {
  name: string;
  description: string,
  abv: number;
  ibu: number;
  uuid: string;
  api_id: string;
  name_of_api: string;
  created_at: Date;
  rating?: Array<BeerRating>;
}
