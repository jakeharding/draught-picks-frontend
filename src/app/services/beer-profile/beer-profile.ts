import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment as Env } from '../../../environments/environment';
import PageResponse from '../../models/PageResponse';
import BeerProfile from '../../models/BeerProfile';

/**
 * Generated class for the BeerProfileProvider provider.
 *
 * See https://angular.io/guide/dependency-injection for more info on providers
 * and Angular DI.
 */
@Injectable({
  providedIn: 'root',
})
export class BeerProfileProvider {
  url: string;

  constructor(public http: HttpClient) {
    this.url = `${Env.REST_API_ROOT}/preferences`;
  }

  /**
   * retrieve function
   * No Parameters
   * Gets the beer profile and returns the result
   */
  retrieve(): Promise<BeerProfile> {
    return this.http.get(this.url).toPromise().then(({ results }: PageResponse<BeerProfile>) => {
      return results[0];
    });
  }

  /**
   * save function
   * Returns Beer Profile
   */
  save(profile: BeerProfile) {
    if (profile.uuid && profile.uuid.length > 1) {
      // If profile has uuid it already exists in DB
      return this.http.put(`${this.url}/${profile.uuid}`, profile).toPromise();
    } else {
      // Create it
      return this.http.post(this.url, profile).toPromise();
    }
  }
}
