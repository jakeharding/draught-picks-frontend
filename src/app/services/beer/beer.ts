import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment as Env } from '../../../environments/environment';
import Beer from '../../models/Beer';
import PageResponse from '../../models/PageResponse';
import { LIMIT } from '../../directives/infinite-scroller/infinite-scroller';
import { EMPTY, Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { AutoCompleteService } from 'ionic4-auto-complete';
/*
  Generated class for the BeerProvider provider.

  See https://angular.io/guide/dependency-injection for more info on providers
  and Angular DI.
*/
@Injectable({
  providedIn: 'root',
})
export class BeerProvider implements AutoCompleteService {
  url: string;
  recentsUrl: string;
  recommendedUrl: string;
  labelAttribute = 'name'; // Attribute needed for the AutoCompleteService
  constructor(public http: HttpClient) {
    this.url = `${Env.REST_API_ROOT}/beers`;
    this.recentsUrl = `${Env.REST_API_ROOT}/recent-beers`;
    this.recommendedUrl = `${Env.REST_API_ROOT}/recommended-beers`;
  }

  /**
   * search function
   * Parameters: Observable of type Beer list
   * Sets the limit and offset for the infinite scrolling
   */
  search(params: any): Observable<Beer[]> {
    if (!params.limit) {
      params.limit = LIMIT;
    }
    if (!params.offset) {
      params.offset = 0;
    }
    return this.http.get(this.url, {params}).pipe(map(({results}: PageResponse<Beer>) => results));
  }

  /**
   * Method called in auto complete search bar.
   * @param beerName - Name of beer
   * @returns - List of beers
   */
  getResults(beerName: string) {
    if (beerName) {
      return this.search({search: beerName});
    }
    return EMPTY;
  }

  /**
   * recents function
   * Parameters: Observable of type Beer list
   * Sets the limit and offset for the infinite scrolling
   */
  recents(params: any): Observable<Beer[]> {
    if (!params.limit) {
      params.limit = LIMIT;
    }
    if (!params.offset) {
      params.offset = 0;
    }

    return this.http.get(this.recentsUrl, {params}).pipe(map(({results}: PageResponse<Beer>) => results));
  }

  /**
   * recommended function
   * Parameters: query params for pagination
   * Retrieves a page of recommended beers for an authenticated user.
   */
  recommended(params: any): Observable<Beer[]> {
    if (!params.limit) {
      params.limit = LIMIT;
    }
    if (!params.offset) {
      params.offset = 0;
    }
    return this.http.get(this.recommendedUrl, {params})
      .pipe(map(({results}: PageResponse<Beer>) => results));
  }

  /**
   * createRecent function
   * Parameters: Beer
   * returns the recent beer list
   */
  createRecent(recent: Beer) {
    return this.http.post(this.recentsUrl, {beer: recent.uuid}).toPromise();
  }

  /**
   * Retrieve a information about a beer.
   * @param uuid - ID of beer
   * @returns - Promise a beer
   */
  retrieve(uuid: string) {
    return this.http.get(`${this.url}/${uuid}`).toPromise();
  }


}
