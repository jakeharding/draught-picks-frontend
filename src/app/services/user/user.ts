import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment as Env } from '../../../environments/environment';
import PageResponse from '../../models/PageResponse';
import User from '../../models/User';
import AuthResponse from '../../models/AuthResponse';
import BeerRating from '../../models/BeerRating';
import Beer from '../../models/Beer';

/**
 * Generated class for the UserProvider provider.
 *
 * See https://angular.io/guide/dependency-injection for more info on providers
 * and Angular DI.
 */
@Injectable({
  providedIn: 'root',
})
export class UserProvider {
  private readonly url: string;
  private readonly resendConfirmEmailUrl: string;
  private readonly confirmEmailUrl: string;
  private readonly passwordResetUrl: string;

  constructor(public http: HttpClient) {
    this.url = `${Env.REST_API_ROOT}/users`;
    this.resendConfirmEmailUrl = `${this.url}/resend-confirm-email`;
    this.confirmEmailUrl = `${this.url}/confirm-email`;
    this.passwordResetUrl = `${this.url}/reset-email`;
  }

  create(userData: AuthResponse | Beer | BeerRating | User) {
    return this.http.post(this.url, userData).toPromise();
  }

  retrieve() {
    return this.http.get(this.url).toPromise().then(({ results }: PageResponse<User>) => {
      return results[0];
    });
  }

  update(user: User): Promise<any> {
    return this.http.put(`${this.url}/${user.uuid}`, user).toPromise();
  }

  resendConfirmEmail({email}): Promise<object> {
    return this.http.post(this.resendConfirmEmailUrl, {email}).toPromise();
  }

  confirmEmail({key}): Promise<any> {
    return this.http.put(this.confirmEmailUrl, {confirm_key: key}).toPromise();
  }

  sendPasswordResetEmail(emailAddress: string): Promise<any> {
    return this.http.post(this.passwordResetUrl, {email: emailAddress}).toPromise();
  }
}
