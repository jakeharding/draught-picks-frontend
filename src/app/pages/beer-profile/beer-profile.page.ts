import { Component, OnInit } from '@angular/core';
import Beer from '../../models/Beer';
import User from '../../models/User';
import BeerProfile from '../../models/BeerProfile';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { NavController } from '@ionic/angular';
import { ToastProvider } from '../../services/toast/toast';
import { BeerProvider } from '../../services/beer/beer';
import { UserProvider } from '../../services/user/user';
import { BeerProfileProvider } from '../../services/beer-profile/beer-profile';
import { BasePage } from '../BasePage';

@Component({
  selector: 'app-beer-profile',
  templateUrl: './beer-profile.page.html',
  styleUrls: ['./beer-profile.page.scss'],
})
// TODO Support more than one profile
export class BeerProfilePage extends BasePage {
  beerResults: Array<Beer>;
  beerSearch: string;
  user: User;
  beerProfileForm: FormGroup;

  constructor(public navCtrl: NavController,
              public toastProvider: ToastProvider, public beerProvider: BeerProvider,
              private userProvider: UserProvider, private profileProvider: BeerProfileProvider,
              public formBuilder: FormBuilder) {
    super('beer-profile');
    this.beerSearch = '';
    this.userProvider.retrieve().then( (user: User) => {
      this.user = user;
    });

    this.beerProfileForm = this.formBuilder.group({
      abv_low: ['', [Validators.pattern('\\d+')]],
      abv_hi: ['', [Validators.pattern('\\d+')]],
      ibu_low: ['', [Validators.pattern('\\d+')]],
      ibu_hi: ['', [Validators.pattern('\\d+')]],
      like_description: ['', []],
      user: [],
      uuid: []
    });
    // TODO Beer profile would be better stored in state
    this.profileProvider.retrieve().then((profile: BeerProfile) => {
      this.beerProfileForm.patchValue(profile || {});
    });
  }

  /**
   * saveProfile function
   * No Parameters
   * Gets the user beer-profile and saves them in the database
   */
  public async saveProfile() {
    this.toastProvider.successToast('Your recommendations will be here soon!');

    this.profileProvider.save(
      Object.assign({}, this.beerProfileForm.value, {user: this.user.uuid})).then((profile: BeerProfile) => {
      this.beerProfileForm.patchValue(profile);
      this.toastProvider.successToast('Your recommendations have arrived! Go to the home tab to see them!');
    }, () => {
      this.toastProvider.errorToast('We couldn\'t save your profile at the moment. Please try again.');
    });

  }

  /**
   * Called when a user selects a favorite.
   * @param beer - Beer object
   */
  favoriteSelected(beer: Beer) {
    if (this.user.favorite_beers.indexOf(beer) < 0) {
      this.user.favorite_beers.push(beer);
      this.userProvider.update(this.user).then((user: User) => {
        this.user.favorite_beers = user.favorite_beers;
      });
    }
    this.beerSearch = '';
  }

  /**
   * Called when a user removes a beer from their favorites.
   * @param beer - beer object
   */
  removeFavorite(beer: Beer) {
    this.user.favorite_beers = this.user.favorite_beers.filter((b: Beer) => {
      return b.uuid !== beer.uuid;
    });
    this.userProvider.update(this.user);
  }

  /**
   * Called in the autocomplete in response to entry
   * @param event - Event object
   */
  search(event: Event) {
    if (this.beerSearch && this.beerSearch.length > 0) {
      this.beerProvider.search(this.beerSearch).toPromise().then((results: Array<Beer>) => {
        // TODO: This would be better used as an observable.
        this.beerResults = results;
      });
    }
  }
}
