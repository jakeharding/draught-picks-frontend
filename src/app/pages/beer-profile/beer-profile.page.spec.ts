/**
 * preferences.spec.ts
 *
 * Created by jake
 * Created on 11/16/18
 *
 * Test the BeerProfilePage.
 */
import { BeerProfilePage } from './beer-profile.page';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { NavController } from '@ionic/angular';
import { BeerProvider } from '../../services/beer/beer';
import { UserProvider } from '../../services/user/user';
import { FormBuilder, FormGroupDirective } from '@angular/forms';
import User from '../../models/User';
import BeerProfile from '../../models/BeerProfile';
import Beer from '../../models/Beer';
import { of } from 'rxjs';
import { mockToastProvider } from '../../../../setup-jest';
import { ToastProvider } from '../../services/toast/toast';
import { BeerProfileProvider } from '../../services/beer-profile/beer-profile';

describe('BeerProfilePage', () => {
  let component: BeerProfilePage;
  let fixture: ComponentFixture<BeerProfilePage>;

  const mockBeer = {uuid: 'beerId'} as Beer;
  const mockUser = {favorite_beers: [mockBeer]} as User;
  const mockUserProvider = {
    retrieve: jest.fn(() => Promise.resolve({} as User)),
    update: jest.fn(() => Promise.resolve(mockUser))
  };
  const mockProfileProvider = {
    retrieve: jest.fn(() => Promise.resolve({} as BeerProfile)),
    save: jest.fn(() => Promise.resolve())
  };

  const mockBeerProvider = {
    search: jest.fn()
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      declarations: [BeerProfilePage, FormGroupDirective],
      providers: [
        {provide: BeerProvider, useValue: mockBeerProvider},
        {provide: UserProvider, useValue: mockUserProvider},
        {provide: BeerProfileProvider, useValue: mockProfileProvider},
        {provide: NavController, useValue: {}},
        {provide: ToastProvider, useValue: mockToastProvider},
        FormBuilder
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(BeerProfilePage);
    component = fixture.debugElement.componentInstance;
  });

  it('should initialize the BeerProfilePage', () => {
    expect(component).toBeDefined();
    expect(fixture).toMatchSnapshot();
  });

  test('properties are set during initialization of component', () => {
    expect(mockUserProvider.retrieve).toHaveBeenCalledTimes(1);
    expect(mockProfileProvider.retrieve).toHaveBeenCalledTimes(1);
    expect(component.user).toBeDefined();
    expect(component.beerProfileForm).toBeDefined();
  });

  test('savePrefs should call the toastController.create, prefsProvider.save, and set the prefsForm values to null',
    async () => {
      await component.saveProfile();
      expect(mockToastProvider.successToast).toHaveBeenCalledTimes(2);
      expect(mockProfileProvider.save).toHaveBeenCalledTimes(1);
      expect(component.beerProfileForm.value.abv_low).toBeNull();
      expect(component.beerProfileForm.value.abv_hi).toBeNull();
      expect(component.beerProfileForm.value.ibu_low).toBeNull();
      expect(component.beerProfileForm.value.ibu_hi).toBeNull();
    });

  test('savePrefs should call the mockToastProvider.successToast, prefsProvider.save, and not change the prefsForm values',
    async () => {
      component.beerProfileForm.value.abv_low = '1';
      component.beerProfileForm.value.abv_hi = '10';
      component.beerProfileForm.value.ibu_low = '10';
      component.beerProfileForm.value.ibu_hi = '10';
      await component.saveProfile();
      expect(mockToastProvider.successToast).toHaveBeenCalledTimes(2);
      expect(mockProfileProvider.save).toHaveBeenCalledTimes(1);
      expect(component.beerProfileForm.value.abv_low).toBe('1');
      expect(component.beerProfileForm.value.abv_hi).toBe('10');
      expect(component.beerProfileForm.value.ibu_low).toBe('10');
      expect(component.beerProfileForm.value.ibu_hi).toBe('10');
    });

  test('savePrefs should display an error toast when prefsProvider.save is rejected', async () => {
    mockProfileProvider.save.mockImplementation(() => Promise.reject('error'));
    await component.saveProfile();
    expect(mockToastProvider.successToast).toHaveBeenCalledTimes(1);
    expect(mockToastProvider.errorToast).toHaveBeenCalledTimes(1);
    expect(mockProfileProvider.save).toHaveBeenCalledTimes(1);
  });

  test('favoriteSelected should call userProvider.update with the new beer in the user\'s favorite', async () => {
    component.user = {favorite_beers: [] as Array<Beer>} as User;
    await component.favoriteSelected(mockBeer);
    expect(mockUserProvider.update).toHaveBeenCalledTimes(1);
    expect(component.user.favorite_beers).toContain(mockBeer);
    expect(component.beerSearch).toBe('');
  });

  test('favoriteSelected should only reset the beerSearch if the beer is already a favorite', () => {
    component.user = mockUser;
    component.favoriteSelected(mockBeer);
    expect(mockUserProvider.update).not.toHaveBeenCalled();
    expect(component.beerSearch).toBe('');
  });

  test('removeFavorite should call remove the beer from favorite beers and call userProvider.update', () => {
    component.user.favorite_beers = [mockBeer];
    component.removeFavorite(mockBeer);
    expect(component.user.favorite_beers).not.toContain(mockBeer);
    expect(mockUserProvider.update).toHaveBeenCalledTimes(1);
  });

  test('search should call beerProvider.search when a value is entered', async () => {
    const mockSearchResult = of([mockBeer]);
    mockBeerProvider.search.mockReturnValue(mockSearchResult);
    component.beerSearch = 'aBeer';
    await component.search({} as Event);
    expect(mockBeerProvider.search).toHaveBeenCalledTimes(1);
    expect(mockBeerProvider.search).toHaveBeenCalledWith(component.beerSearch);

  });
});
