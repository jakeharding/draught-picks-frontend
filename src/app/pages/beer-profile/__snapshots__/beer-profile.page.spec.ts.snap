// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeerProfilePage should initialize the BeerProfilePage 1`] = `
<app-beer-profile
  beerProfile={[Function Object]}
  beerProfileForm={[Function FormGroup]}
  beerProvider={[Function Object]}
  beerSearch=""
  formBuilder={[Function FormBuilder]}
  navCtrl={[Function Object]}
  profileProvider={[Function Object]}
  toastProvider={[Function Object]}
  user={[Function Object]}
  userProvider={[Function Object]}
>
  <ion-header>
    <draught-header />
  </ion-header><ion-content
    class="ion-padding"
  >
    <ion-row>
      <ion-col
        class="ion-text-center"
        offsetlg="3"
        offsetmd="3"
        offsetsm="1"
        offsetxl="4"
        size="12"
        sizelg="6"
        sizemd="6"
        sizesm="10"
        sizexl="4"
      >
        <h1>
          Tell us about the beer you like!
        </h1>
        <p>
          <i>
            The more you tell us the better we can recommend beers.
          </i>
        </p>
        <ion-row>
          <h3>
            Select some of your favorites:
          </h3>
          <ion-auto-complete
            class="ion-text-left"
            tid="enterBeerInput"
          />
        </ion-row>
        
        <h2>
          Ranges of ABV and IBU
        </h2>
        <ion-row
          class="ion-text-center"
        >
          <ion-col>
            <h3>
              ABV 
              <ion-icon
                class="infoButton"
                color="primary"
                name="information-circle"
                routerdirection="forward"
                routerlink="./abv-info"
                slot="end"
                tid="abvInfoButton"
              />
            </h3>
          </ion-col>
          <ion-col>
            <h3>
              IBU 
              <ion-icon
                class="infoButton"
                color="primary"
                name="information-circle"
                routerdirection="forward"
                routerlink="./ibu-info"
                tid="ibuInfoButton"
              />
            </h3>
          </ion-col>
        </ion-row>
        <form>
          <ion-row
            class="ion-text-center"
          >
            <ion-col
              class="ion-text-center"
            >
              <ion-item>
                <ion-label
                  color="primary"
                  position="floating"
                >
                  High
                </ion-label>
                <ion-input
                  formcontrolname="abv_hi"
                  tid="enterAbvHigh"
                  type="number"
                />
              </ion-item>
              <ion-item>
                <ion-label
                  color="primary"
                  position="floating"
                >
                  Low
                </ion-label>
                <ion-input
                  formcontrolname="abv_low"
                  tid="enterAbvLow"
                  type="number"
                />
              </ion-item>
            </ion-col>
            <ion-col
              class="ion-text-center"
            >
              <ion-item>
                <ion-label
                  color="primary"
                  position="floating"
                >
                  High
                </ion-label>
                <ion-input
                  formcontrolname="ibu_hi"
                  tid="enterIbuHigh"
                  type="number"
                />
              </ion-item>
              <ion-item>
                <ion-label
                  color="primary"
                  position="floating"
                >
                  Low
                </ion-label>
                <ion-input
                  formcontrolname="ibu_low"
                  tid="enterIbuLow"
                  type="number"
                />
              </ion-item>
            </ion-col>
          </ion-row>
          <ion-row
            class="ion-margin"
          >
            <h2>
              Describe your favorite beer(s) below: 
              <ion-icon
                class="infoButton"
                color="primary"
                name="information-circle"
                routerlink="./beer-favorite-info"
                tid="descInfoButton"
              />
            </h2>
          </ion-row>
          <ion-row>
            <ion-textarea
              formcontrolname="like_description"
              rows="7"
              tid="enterTextArea"
            />
          </ion-row>
          <ion-button
            tid="submitBtn"
          >
             Submit 
          </ion-button>
        </form>
      </ion-col>
    </ion-row>
  </ion-content>
</app-beer-profile>
`;
